build_tools/cmake/iree_check_test.cmake

```diff
@@ -0,0 +1,153 @@
+  # reach into the internals.
+  set(_GENERATED_MODULE_NAME "${_MODULE_TARGET_NAME}.module")
+  add_custom_target(
+    "${_NAME}" ALL
+     DEPENDS
+       "${_GENERATED_MODULE_NAME}"
+  )
+
+  string(REPLACE "_" "/" _PACKAGE_PATH ${_PACKAGE_NAME})                                                                                                    
+  set(_NAME_PATH "${_PACKAGE_PATH}:${_RULE_NAME}") 
+
+  add_test(
+    NAME
+      "${_NAME_PATH}"
+    COMMAND
+      "${CMAKE_SOURCE_DIR}/build_tools/cmake/run_test.${IREE_HOST_SCRIPT_EXT}"
+      "$<TARGET_FILE:iree_modules_check_iree-check-module>"
+      "--input_file=${CMAKE_CURRENT_BINARY_DIR}/${_GENERATED_MODULE_NAME}"
+      "--driver=${_RULE_DRIVER}"
+      "${_RULE_ARGS}"
+  )
+
+  set_property(TEST "${_NAME_PATH}" PROPERTY REQUIRED_FILES "${_GENERATED_MODULE_NAME}")
+  set_property(TEST "${_NAME_PATH}" PROPERTY ENVIRONMENT "TEST_TMPDIR=${_NAME}_test_tmpdir")
+  set_property(TEST "${_NAME_PATH}" PROPERTY LABELS "${_PACKAGE_PATH}")
+endfunction()
+
+
+# iree_check_test_suite()
+#
+# Creates a test suite of iree-check-module tests.
+#
+# Mirrors the bzl rule of the same name.
+#
+# One test is generated per source and backend/driver pair.
+# Parameters:
+#   NAME: name of the generated test suite.
+#   SRCS: source mlir files containing the module.
+#   TARGET_BACKENDS: backends to compile the module for. These form pairs with
+#       the DRIVERS argument (due to cmake limitations they are separate list
+#       arguments). The lengths must exactly match. If no backends or drivers are
+#       specified, a test will be generated for every supported pair.
+#   DRIVERS: drivers to run the module with. These form pairs with the
+#       TARGET_BACKENDS argument (due to cmake limitations they are separate list
+#       arguments). The lengths must exactly match. If no backends or drivers are
+#       specified, a test will be generated for every supported pair.
+#   ARGS: additional args to pass to the underlying iree-check-module tests. The
+#       driver and input file are passed automatically. To use different args per
+#       test, create a separate suite or iree_check_test.
+function(iree_check_test_suite)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME"
+    "SRCS;TARGET_BACKENDS;DRIVERS;ARGS"
+    ${ARGN}
+  )
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  if(NOT DEFINED _RULE_TARGET_BACKENDS AND NOT DEFINED _RULE_DRIVERS)
+    set(_RULE_TARGET_BACKENDS "vmla" "vulkan-spirv")
+    set(_RULE_DRIVERS "vmla" "vulkan")
+  endif()
+
+  list(LENGTH _RULE_TARGET_BACKENDS _TARGET_BACKEND_COUNT)
+  list(LENGTH _RULE_DRIVERS _DRIVER_COUNT)
+
+  if(NOT _TARGET_BACKEND_COUNT EQUAL _DRIVER_COUNT)
+    message(SEND_ERROR
+        "TARGET_BACKENDS count ${_TARGET_BACKEND_COUNT} does not match DRIVERS count ${_DRIVER_COUNT}")
+  endif()
+
+  math(EXPR _MAX_INDEX "${_TARGET_BACKEND_COUNT} - 1")
+  foreach(_INDEX RANGE "${_MAX_INDEX}")
```

Comment:
Oh CMake math magic :see_no_evil:

Reply:
Yeah this was the most disgusting list iteration I have ever written :-P

Reply:
That is what I assumed and therefore needed to comment ;)